### PR TITLE
Improving SUT and test suite stability

### PR DESCRIPTION
**Pull request by augustocristian**
Several issues causing test flakiness were addressed:
1. Problems related to the number of inotify instances.
2. Concurrency and lockdown issues with proxies, solved using custom Dockerfiles.
3. Dependency and SQL problems caused by the system starting under heavy load without migrations or a ready database.
4. Issues with order states remaining in a pending validation status under certain system loads were resolved.

Several minor fixes:
- Enhanced integration of the logging system (RETORCH tool) within the Jenkinsfile.
- Improved the readability of compose files, scripts, and code.

### CODE REVIEW

**Comment by javiertuya** [comment]
@augustocristian Check build 45 on this branch

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 1]
```diff
@@ -76,6 +80,16 @@ void setup(TestInfo testInfo) {
...
+        log.debug("Waiting for the SUT ready on {}.", sutUrl);
+        Wait<WebDriver> waitSUTReady = new FluentWait<>(this.driver)
+                .withTimeout(Duration.ofSeconds(30))
+                .pollingEvery(Duration.ofSeconds(5))
+                .ignoring(NoSuchElementException.class);
+        waitSUTReady.until((webDriver) -> {
+            webDriver.navigate().refresh();
+            return driver.findElement(By.xpath("//img"));
+        });
+
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:92]: Shouldn't this be in BeforeAll?, maybe enclosed un a separate method

File: retorchfiles/scripts/waitforSUT.sh [hunk 2]
```diff
@@ -1,19 +1,45 @@
...
+  local TJOB_NAME="$1"
+  local DB_SERVICE_NAME="sqldata_$TJOB_NAME"
+  local WEB_SERVICE_URL="http://webmvc_$TJOB_NAME:80"
+  local DB_NAME="[Microsoft.eShopOnContainers.Services.CatalogDb]"
+  local TABLE_NAME="Catalog"
+  local QUERY="USE $DB_NAME; SELECT COUNT (*) FROM $TABLE_NAME;"
+
+  while true; do
+    # Check if frontend service is up
+    if curl --insecure -s "$WEB_SERVICE_URL" | grep -q "<div class=\"esh-catalog-item col-md-4\">"; then
```
Comment (javiertuya) [inline retorchfiles/scripts/waitforSUT.sh:16]: As you have included a wait for the sut UI in the tests, is this redundant now?

File: retorchfiles/scripts/waitforSUT.sh [hunk 3]
```diff
@@ -1,19 +1,45 @@
...
+      # Check if database service is up
+      if docker exec "$DB_SERVICE_NAME" /opt/mssql-tools/bin/sqlcmd -S localhost -U SA -P 'Pass@word' -Q "$QUERY" | grep -q "14"; then
```
Comment (javiertuya) [inline retorchfiles/scripts/waitforSUT.sh:18]: Do not hardoce passwords nor include magic numbers that belong to the application. At the end, is this going to be checked in the tests?

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: retorchfiles/scripts/waitforSUT.sh (hunk 2 above)
Comment (augustocristian) [reply retorchfiles/scripts/waitforSUT.sh:16]: @javiertuya 
I added a wait script to my tests but I realized that with it the time measurements are incorrect . However, the code in the test is only temporary. I am currently working on a more comprehensive database waiter in the Set-up @beforall method that creates a connection to the DB and checks that everything is loaded before starting the test.

Comment (javiertuya) [reply retorchfiles/scripts/waitforSUT.sh:16]: If so, then remove from the tests to allow finish this PR. The priority is to have a stable system, there is time later for enhancements.

**Comment by augustocristian** [comment]
@javiertuya 
Several changes, including fixes to various versions of Docker, Compose, and containers, which seem to have resolved the test flakiness in TJobC. However, we are now finding some flakiness in TJobE attributed to timeouts and failure to update the order state. I believe this issue should be tackled in a separate branch.

Additionally, I've addressed proxy issues by implementing a combination of solutions suggested by Javier. This includes employing a custom waiter in the test method to monitor the database state, along with creating custom images tailored to each scenario's configuration

**Comment by javiertuya** [comment]
@augustocristian @ClaudiodelaRiva Remember: no changes until all critical regression issues are done.

In this case it seems that issue #56 has been already fixed without this PR, but not closed. Am I right?

**Comment by augustocristian** [comment]
@javiertuya  I've resolved all the flakiness issues in the system under test (SUT) caused by parallelization and low resource availability!

File: pom.xml [hunk 4]
```diff
@@ -69,6 +69,11 @@
 				<artifactId>retorch-annotations</artifactId>
 				<version>${retorch-annotations.version}</version>
 			</dependency>
+			<dependency>
+				<groupId>com.microsoft.sqlserver</groupId>
+				<artifactId>mssql-jdbc</artifactId>
+				<version>12.7.0.jre11-preview</version>
```
Comment (javiertuya) [inline pom.xml:75]: Use only stable versions

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 5]
```diff
@@ -168,4 +172,106 @@ protected void logout() throws ElementNotFoundException {
...
+    protected static void checkDBMigration(){
+
+        // Get properties
+        String user = properties.getProperty("SQLDB_USER");
+        String password = properties.getProperty("SQLDB_PASSWORD");
+        String host = "sqldata_" + tJobName;
+        //host = "localhost"; // default host
+
+        // Build JDBC URL
+        final int MAX_TABLES = 6;
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:185]: Why six tables are the max? Justify

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 6]
```diff
@@ -168,4 +172,106 @@ protected void logout() throws ElementNotFoundException {
...
+        String url = "jdbc:sqlserver://" + host + ":1433;Encrypt=True;TrustServerCertificate=True;user=" + user + ";password=" + password;
+
+        int iter = 0;
+        boolean found = false;
+        while (!found && iter < MAX_ITERATIONS) {
+            iter++;
+            int numTables = 0;
+            try (Connection connection = DriverManager.getConnection(url)) {
+                try (PreparedStatement stmt = connection.prepareStatement(query)) {
+                    try (ResultSet rs = stmt.executeQuery()) {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:199]: Use a single try with resources

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 7]
```diff
@@ -168,4 +172,106 @@ protected void logout() throws ElementNotFoundException {
...
+                try {
+                    Thread.sleep(WAIT_TIME_MS);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+
+        if (!found) {
+            log.error("The databases are not migrated after " + MAX_ITERATIONS + " attempts.");
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:223]: This is a non-recoverable situation, test should not run beyond this point

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 8]
```diff
@@ -168,4 +172,106 @@ protected void logout() throws ElementNotFoundException {
...
+        }
+
+
+    }
+    protected static void checkCatalogDBStatus() {
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:228]: Blank lines do not separate blocks

File: src/test/java/giis/eshopcontainers/e2e/functional/tests/OrderTests.java [hunk 9]
```diff
@@ -90,19 +90,34 @@ void testCancelOrder() throws ElementNotFoundException {
...
+
+            WebElement lastOrder = listOrders.get(initialNOrders);
+            WebElement statusElement = lastOrder.findElements(By.className("esh-orders-item")).get(3);
 
-        Assertions.assertEquals(initialNOrders + 1, listOrders.size(), "Order count is not as expected");
+            actualState = statusElement.getText();
+            log.debug("End of iteration {}, the order state is {}", iter, actualState);
+
+            if (!actualState.equals("awaitingvalidation")) {
+                break; // Exit loop if the state is not awaiting validation
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/tests/OrderTests.java:113]: This comment only says what we can infer from the code, but not why

File: sut/src/ApiGateways/Envoy/config/mobileshopping/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,9 @@
+FROM envoyproxy/envoy:v1.11.1
+
+ARG TAG
```
Comment (javiertuya) [inline sut/src/ApiGateways/Envoy/config/mobileshopping/Dockerfile:3]: Use blank lines to separate logical blocks, not for separate EVERY line

File: sut/sutdebugging/Jenkinsfile-tjobc [hunk 11]
```diff
@@ -0,0 +1,117 @@
+pipeline {
```
Comment (javiertuya) [inline sut/sutdebugging/Jenkinsfile-tjobc:1]: Is this pipeline temporary and then, should it be deleted now?

**Review by javiertuya — CHANGES_REQUESTED** [review]
Review title and description of the PR to reflect what it is doing

Some review comments are applicable to other places too

**Comment by javiertuya** [comment]
@augustocristian Do not add anything new to this endless PR, just revise the review comments. Any flakiness after merging the PR can be addressed later in separate issues/PRs

FYI @ClaudiodelaRiva

**Comment by augustocristian** [comment]
So, this PR is ready for review and merge @javiertuya

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java (hunk 7 above)
Comment (javiertuya) [reply src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:218]: Don't log and throw, just fail the test (this is junit)

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java [hunk 12]
```diff
@@ -173,15 +175,15 @@ protected void logout() throws ElementNotFoundException {
...
-    protected static void checkDBMigration() throws SQLDataException {
+    protected static void checkDBMigration() {
         // Get properties
         String user = properties.getProperty("SQLDB_USER");
         String password = properties.getProperty("SQLDB_PASSWORD");
         String host = "sqldata_" + tJobName;
-        // Build JDBC URL
-        final int MAX_TABLES = 6;
+        // Minimal amount of tables expected in the database
+        final int MIN_TABLES = 6;
```
Comment (javiertuya) [inline src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java:186]: The comment is not explaining why. Another comment before mentions seven databases, however, here, the variable says six, and it refers to tables, not databases.

**Review by javiertuya — CHANGES_REQUESTED** [review]

File: src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java (hunk 12 above)
Comment (augustocristian) [reply src/test/java/giis/eshopcontainers/e2e/functional/common/BaseLoggedClass.java]: I've aligned the two comments and the conditional @javiertuya

**Review by javiertuya — APPROVED** [review]
Merging, but keep in mind for future reviews:
- When addressing a change in a line, check the context around it, for instance, in this case you are comparing `numTables >= MIN_DATABASES`, the name pf the variable is inconsistent.
- Better do not duplicate comments, here the inline comment is nearly the same than the last statement in the method comment.

**Pull request merged**